cmd/promtool/debug.go

```diff
@@ -0,0 +1,133 @@
+// Copyright 2018 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"bytes"
+	"fmt"
+	"net/http"
+	"os"
+
+	"github.com/google/pprof/profile"
+)
+
+type debugWriterConfig struct {
+	server         string
+	tarballName    string
+	pathToFileName map[string]string
+	postProcess    func(b []byte) (*bytes.Buffer, error)
+}
+
+type debugWriter struct {
+	archiver
+	httpClient
+	requestToFile map[*http.Request]string
+	postProcess   func(b []byte) (*bytes.Buffer, error)
+}
+
+func newDebugWriter(cfg debugWriterConfig) *debugWriter {
+	client, err := newHTTPClient(httpClientConfig{serverURL: cfg.server})
+	if err != nil {
+		panic(err)
+	}
+	archiver, err := newArchiver(archiverConfig{archiveName: cfg.tarballName})
+	if err != nil {
+		panic(err)
+	}
+	reqs := make(map[*http.Request]string)
+	for path, filename := range cfg.pathToFileName {
+		req, err := http.NewRequest(http.MethodGet, client.urlJoin(path), nil)
+		if err != nil {
+			panic(err)
+		}
+		reqs[req] = filename
+	}
+	return &debugWriter{
+		archiver,
+		client,
+		reqs,
+		cfg.postProcess,
+	}
+}
+
+func (w *debugWriter) Write() int {
+	for req, filename := range w.requestToFile {
+		_, body, err := w.do(req)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			return 1
+		}
+
+		buf, err := w.postProcess(body)
+		if err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			return 1
+		}
+
+		if err := w.archiver.Write(filename, buf); err != nil {
+			fmt.Fprintln(os.Stderr, err)
+			return 1
+		}
+	}
+
+	if err := w.Close(); err != nil {
+		fmt.Fprintln(os.Stderr, err)
+		return 1
+	}
+
+	fmt.Println("debug complete all files written in:", w.File().Name())
+	return 0
+}
+
+func validate(b []byte) (*profile.Profile, error) {
+	p, err := profile.Parse(bytes.NewReader(b))
+	if err != nil {
+		return nil, err
+	}
+	return p, nil
+}
+
+func buffer(p *profile.Profile) (*bytes.Buffer, error) {
+	buf := new(bytes.Buffer)
```

Comment:
Since the return type is `*bytes.Buffer` so I decided to use `new` for getting a pointer. These two approaches both seem to be recommended by the [official document](https://golang.org/pkg/bytes/#Buffer). Actually I was bothered by the case when it occurs error. What should I return as the empty value of `bytes.Buffer`? And I feel it's simpler to just use a pointer over than a value so `nil` can be the empty value. What do you think?

Reply:
If you expand the first example for `bytes.Buffer` (https://golang.org/pkg/bytes/#Buffer), it also just does `var b bytes.Buffer` and then takes the address of it when it's needed (`&b`). But yeah, it doesn't matter too much.

You don't seem to be actually needing any of the `bytes.Buffer` functionality (like writing more into it) after this function returns, so I would just return `buf.Bytes()` directly as a `[]byte` slice.